Source/WebCore/platform/graphics/WidthIterator.cpp

```diff
@@ -230,82 +206,240 @@ static void addToGlyphBuffer(GlyphBuffer& glyphBuffer, Glyph glyph, const Font&
         glyphBuffer.add(0, font, 0, currentCharacterIndex + 1);
 }
 
+struct SmallCapsState {
+    const Font* lastFont;
+    const Font* font;
+    const Font* synthesizedFont { nullptr };
+    const Font* smallSynthesizedFont { nullptr };
+    bool isSmallCaps { false };
+    bool isLastSmallCaps { false };
+    bool shouldSynthesizeCharacter { false };
+
+    FontVariantCaps fontVariantCaps { FontVariantCaps::Normal };
+    bool dontSynthesizeSmallCaps { true };
+    bool engageAllSmallCapsProcessing { false };
+
+    void setSmallCapsData(const Font* font, bool isSmallCaps, const FontDescription& fontDescription)
+    {
+        ASSERT(font);
+        this->lastFont = this->font;
+        this->font = font;
+        synthesizedFont = &font->noSynthesizableFeaturesFont();
+        smallSynthesizedFont = synthesizedFont->smallCapsFont(fontDescription);
+        this->isLastSmallCaps = this->isSmallCaps;
+        this->isSmallCaps = isSmallCaps;
+    }
+    void clear()
+    {
+        synthesizedFont = nullptr;
+        smallSynthesizedFont = nullptr;
+        isSmallCaps = false;
+    }
+    void setIsSmallCaps(bool isSmallCaps)
+    {
+        isLastSmallCaps = this->isSmallCaps;
+        this->isSmallCaps = isSmallCaps;
+    }
+
+    bool skipSmallCapsProcessing()
+    {
+        return fontVariantCaps == FontVariantCaps::Normal;
+    }
+};
+
+struct AdvanceState {
```

Comment:
This is a bit confusing, because the word "advance" has 2 meanings:
- `advanceInternal()`
- each glyph has an advance (e.g `CTFontGetAdvancesForGlyphs()`)

This might be better if it was renamed. Perhaps `StateForAdvanceInternal`? `AdvanceInternalState`? Something like that.